### DISCUSSION TITLE
Showing delivery only if the phone number is valid?

### DISCUSSION THREAD

**Discussion by sairmali** [python-smpplib/python-smpplib#231] [Q&A]
hello team,

When I send a message to a wrong number, it still shows as delivered and returns a reference number. How can I fix this so that it only shows as delivered if the phone number is valid?

**Comment by sairmali** [comment]
My operator sent me this message, but I don't understand it. How can I implement it?


Below are Message IDs NOT error codes .
The SMSID returned from SMSC when the SMS submitted successfully to SMSC , you find it on SUBMIT_SM_RESP. 
Regarding the delivery , you can check the SMS status on Deliver_sm_resp .
Best regards,

**Comment by eigenein** [comment]
Hi @sairmali,

They are correct. From `submit_sm` aka `client.send_message()` a reference message ID ([SMPP spec](https://smpp.org/SMPP_v5.pdf) 4.7.14). At this point, the message is queued, but its delivery is not yet attempted.

If you want to check a delivery status, you have 2 options:

### 1. Listen to `deliver_sm` messages

This is what your operator suggested. You need to set `registered_delivery=True` in `client.send_message()` and an incoming message listener via `client.set_message_received_handler()` – see `README.md` for an example.

The handler will receive an instance of [`DeliverSM`](https://github.com/python-smpplib/python-smpplib/blob/2.2.3/smpplib/command.py#L752) as its only parameter, see also SMPP spec 4.3.1.

You'd need to check [`esm_class`](https://github.com/python-smpplib/python-smpplib/blob/2.2.3/smpplib/command.py#L764), SMPP spec 4.7.12. [`SMPP_MSGTYPE_DELIVERYACK`](https://github.com/python-smpplib/python-smpplib/blob/2.2.3/smpplib/consts.py#L185) would mean this is a delivery receipt.

### 2. Manually issue [`query_sm`](https://github.com/python-smpplib/python-smpplib/blob/2.2.3/smpplib/command.py#L831) command

See also SMPP spec 4.5.2. This command would allow checking the status of the submitted message at any time. Note, that it takes time for SMSC to process a message, so you won't see a new status immediately.

**Reply by sairmali** [reply]
Thanks sir for reply this question

**Comment by sairmali** [comment]
i am using only this script so, how can i pass "deliver_sm"


import logging
import sys

import smpplib.gsm
import smpplib.client
import smpplib.consts

logging.basicConfig(level='DEBUG')

parts, encoding_flag, msg_type_flag = smpplib.gsm.make_parts(u'Привет мир!\n'*10)

client = smpplib.client.Client('example.com', SOMEPORTNUMBER, allow_unknown_opt_params=True)

client.set_message_sent_handler(
    lambda pdu: sys.stdout.write('sent {} {}\n'.format(pdu.sequence, pdu.message_id)))
client.set_message_received_handler(
    lambda pdu: sys.stdout.write('delivered {}\n'.format(pdu.receipted_message_id)))

client.connect()
client.bind_transceiver(system_id='login', password='secret')

for part in parts:
    pdu = client.send_message(
        source_addr_ton=smpplib.consts.SMPP_TON_INTL,
        source_addr='SENDERPHONENUM',
        dest_addr_ton=smpplib.consts.SMPP_TON_INTL,
        destination_addr='PHONENUMBER',
        short_message=part,
        data_coding=encoding_flag,
        esm_class=msg_type_flag,
        registered_delivery=True,
    )
    print(pdu.sequence)
client.listen()

**Reply by sairmali** [reply]
@eigenein please tell me this code is valid ?

**Reply by eigenein** [reply]
This looks like the example from `README.md`, so I'm not completely sure why you're asking

**Reply by sairmali** [reply]
i have implemented like this but not getting result

my question is how can i use "deliver_sm"  ?

**Reply by eigenein** [reply]
For that, please refer to [the answer above](https://github.com/python-smpplib/python-smpplib/discussions/231#discussioncomment-7044675) and [SMPP specification](https://smpp.org/SMPP_v5.pdf), or elaborate what your specific problem is

**Reply by sairmali** [reply]
@eigenein Please can you write a function because i am not able to write this really this is great help

**Reply by eigenein** [reply]
I'm sorry, we are here to maintain the package and advise in using it, but not to tutor one in programming or write code for someone else

**Comment by sairmali** [comment]
Hello @eigenein,

can you tell me how can i use "query_sm"


from smpplib import Client, PDU_TYPE_QUERY_SM

SMSC_HOST = "your_smpp_server_host"
SMSC_PORT = your_smpp_server_port
SYSTEM_ID = "your_system_id"
PASSWORD = "your_password"
MESSAGE_ID = "your_message_id"  # ID of the message you want to query
DELAY_BETWEEN_QUERIES = 5  # Time in seconds between queries (adjust as needed)



client = Client(SMSC_HOST, SMSC_PORT, SYSTEM_ID, PASSWORD)
pdu = client.QuerySm(message_id=MESSAGE_ID)




while True:
    client.send_pdu(pdu)
    response = client.recv_pdu()

    if isinstance(response, PDU_TYPE_QUERY_SM_RESP):
        message_state = response.message_state
        final_delivery_receipt = response.final_delivery_receipt
        error_code = response.error_code

        print(f"Message state: {message_state}")
        print(f"Final delivery receipt: {final_delivery_receipt}")
        print(f"Error code: {error_code}")

        # Implement logic based on the received information (e.g., log status, retry sending, etc.)
        break

    else:
        print(f"Unexpected PDU type received: {type(response)}")
        # Handle unexpected responses
        break

    time.sleep(DELAY_BETWEEN_QUERIES)


i am try to do but unefine getting error
